File: backend/app/api/v1/rates.py
import uuid
from datetime import date
from typing import Annotated

from fastapi import APIRouter, Query, status

from app.api.dependencies import MarketDataServiceDependency
from app.schemas.rate import LoanRateBatchCreate, LoanRateRead

router = APIRouter(prefix="/funds/{fund_id}/rates", tags=["rates"])

FromDateQuery = Annotated[date | None, Query()]
ToDateQuery = Annotated[date | None, Query()]
LimitQuery = Annotated[int | None, Query(ge=1, le=5000)]


@router.post("", response_model=list[LoanRateRead], status_code=status.HTTP_201_CREATED)
def add_rates(
    fund_id: uuid.UUID,
    payload: LoanRateBatchCreate,
    service: MarketDataServiceDependency,
) -> list[LoanRateRead]:
    return [LoanRateRead.model_validate(rate) for rate in service.add_rates(fund_id, payload)]


@router.get("", response_model=list[LoanRateRead])
def list_rates(
    fund_id: uuid.UUID,
    service: MarketDataServiceDependency,
    from_date: FromDateQuery = None,
    to_date: ToDateQuery = None,
    limit: LimitQuery = None,
) -> list[LoanRateRead]:
    rates = service.list_rates(fund_id, from_date=from_date, to_date=to_date, limit=limit)
    return [LoanRateRead.model_validate(rate) for rate in rates]

File: frontend/src/api/rates.ts
import client from "./client";
import type { LoanRate, LoanRateCreate } from "../types/api";

export const fetchRates = (fundId: string): Promise<LoanRate[]> =>
  client.get<LoanRate[]>(`/funds/${fundId}/rates`).then((r) => r.data);

export const addRates = (
  fundId: string,
  items: LoanRateCreate[],
): Promise<LoanRate[]> =>
  client
    .post<LoanRate[]>(`/funds/${fundId}/rates`, { items })
    .then((r) => r.data);
